pandas/tests/io/test_excel.py

```diff
@@ -65,135 +64,89 @@ class ReadingTestsBase(SharedItems):
     # This is based on ExcelWriterBase
 
     @pytest.fixture(autouse=True, params=['xlrd', None])
-    def set_engine(self, request):
-        func_name = "get_exceldf"
-        old_func = getattr(self, func_name)
-        new_func = partial(old_func, engine=request.param)
-        setattr(self, func_name, new_func)
-        yield
-        setattr(self, func_name, old_func)
-
-    def get_csv_refdf(self, basename):
+    def cd_and_set_engine(self, request, datapath, monkeypatch):
         """
-        Obtain the reference data from read_csv with the Python engine.
-
-        Parameters
-        ----------
-
-        basename : str
-            File base name, excluding file extension.
-
-        Returns
-        -------
-
-        dfref : DataFrame
+        Change directory and set engine for read_excel calls.
         """
-        pref = os.path.join(self.dirpath, basename + '.csv')
-        dfref = read_csv(pref, index_col=0, parse_dates=True, engine='python')
-        return dfref
+        func = partial(read_excel, engine=request.param)
 
-    def get_excelfile(self, basename, ext):
-        """
-        Return test data ExcelFile instance.
-
-        Parameters
-        ----------
-
-        basename : str
-            File base name, excluding file extension.
-
-        Returns
-        -------
-
-        excel : io.excel.ExcelFile
-        """
-        return ExcelFile(os.path.join(self.dirpath, basename + ext))
+        with monkeypatch.context() as m:
+            m.chdir(datapath("io", "data"))
+            m.setitem(globals(), 'read_excel', func)
+            yield
```

Comment:
nice!

i'm fairly sure you don't even need the `monkeypatch.context` or `yield` here, that's taken care of with the fixture setup/teardown.

so just
```
monkeypatch.chdir(...)
monkeypatch.setitem(...)
```

from docs... "[monkeypatch.context() only needed if you need to] undo some patches before the test ends, such as mocking stdlib functions that might break pytest itself if mocked"

i also think that `monkeypatch.setattr` is more appropriate. https://docs.pytest.org/en/latest/reference.html#_pytest.monkeypatch.MonkeyPatch.setattr

so you could try something like (i think any should work):
```
monkeypatch.setattr('read_excel', func)
monkeypatch.setattr(pandas.io.excel, 'read_excel', func)
monkeypatch.setattr('pandas.io.excel.read_excel', func)
```

although you might need to replace pandas with pd if that doesn't work

Reply:
Thanks for the tip! w.r.t. to setattr I'll have to mess around with it some more. I patched globals because that's where `read_excel` was referenced from in the module. Might also work to prefix the calls with `pd` and patch that as the target